### PR TITLE
add uri template.

### PR DESCRIPTION
**Pull request by vwxyzh**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        private readonly string _template;
+        private readonly Func<string, T> _func;
+        private readonly IUriTemplatePipeline<T>[] _pipeline;
+        private readonly string[][] _parameters;
+
+        /// <summary>
+        /// step 1 (macro):
+        /// * env var:      {%abc%}                         -> expand env var abc.
+        /// Step 2 (pipeline):
+        /// * pipeline:     |> function parameters          -> "", and append a pipeline.
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: >function parameters [](start = 31, length = 19)

seems function name is not used. What about move parameters into UriTemplatePipeline's ctor? Then step2 isn't needed in template string.

File: src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            var t = _marcoRegex.Replace(
+                template,
+                m => Environment.GetEnvironmentVariable(m.Groups[1].Value));
+            List<IUriTemplatePipeline<T>> pipeline = new List<IUriTemplatePipeline<T>>();
+            List<string[]> parameters = new List<string[]>();
+            t = _pipelineRegex.Replace(
+                t,
+                m =>
+                {
+                    pipeline.Add(pipelineProvider(m.Groups[1].Value));
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: >pipeline.Add [](start = 20, length = 12)

seems the list will contain one item at most. No need to use a list?

File: src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs (hunk 1 above)
Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: Function name is the contract name of IUriTemplatePipeline\<T\>.
If move to parameters into UriTemplatePipeline's ctor, i need one more interface: `IUriTemplatePipelineFactory<T>`

File: src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs (hunk 2 above)
Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: no, you can write multi-pipeline in one uri template, e.g.: `a|>x|>y`

Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: @superyyrrzz , see new test case.

File: src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs (hunk 1 above)
Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: @superyyrrzz , see the create method added in unit test.

File: src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+
+    using Microsoft.DocAsCode.Plugins;
+
+    public class UriTemplate<T>
+    {
+        private static readonly Regex _marcoRegex = new Regex(@"{%\s*([\S]+?)\s*%}");
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: Compiled?

File: src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            _template = template;
+            _func = func;
+            _pipeline = pipeline;
+            _parameters = parameters;
+        }
+
+        /// <summary>
+        /// Evaluate uri template with variables
+        /// </summary>
+        public T Evaluate(IDictionary<string, string> variables)
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: >variables [](start = 54, length = 9)

null check?

Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: >string [](start = 46, length = 6)

can be object?

Comment (vwxyzh) [reply src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs]: no, string only.

**Pull request merged**